[Iceberg] Add support for TRUNCATE

ZacBlanco opened the pull request:
## Description

This commit adds support for deleting records from iceberg tables using the query `TRUNCATE TABLE <table>;`

## Motivation and Context

This allows removing data from Iceberg tables without having to drop the table from the catalog.

## Impact

Additional syntax support.

## Test Plan

Test cases:
- `TRUNCATE TABLE x` with a table that has no rows
- `TRUNCATE TABLE x` on a table with records
- Truncate and insert on the same table multiple times back to back


## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== RELEASE NOTES ==

General Changes
* The Iceberg connector now supports `TRUNCATE TABLE <table>` statements
```

ZacBlanco commented:
I will squash the commits once review is finished

ClarenceThreepwood commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java:
```diff
@@ -264,9 +276,35 @@ public ColumnHandle getUpdateRowIdColumnHandle(ConnectorSession session, Connect
-        throw new PrestoException(NOT_SUPPORTED, "This connector only supports delete where one or more partitions are deleted entirely");
+        IcebergTableLayoutHandle layoutHandle = (IcebergTableLayoutHandle) tableLayoutHandle;
+        if (!layoutHandle.getTupleDomain().equals(TupleDomain.all())) {
+            throw new PrestoException(INVALID_ARGUMENTS, "iceberg metadata delete doesn't support predicates");
```
Is this reachable given the check in the previous function?

ZacBlanco replied:
I couldn't generate a test case where this exception would be thrown. However while I assumed that `metadataDelete` would only be called when `supportsMetadataDelete` returns true, I don't actually know the contract between this method and `supportsMetadataDelete`. Is it ever possible it could change in the future. I added this conditional in as a defensive measure in case this was ever called. I suppose instead I could change this condition to just `if (!supportsMetadataDelete(session, tableHandle, Optional.of(layoutHandle))) {` instead to consolidate the logic in the case we want to keep this defensive check

If you think it is unnecessary, then I am not really against removing it

ClarenceThreepwood replied:
I think it is low overhead and ok to leave it. Yes - consolidating the two checks makes sense

ZacBlanco replied:
fixed

ClarenceThreepwood commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/Partition.java:
```diff
@@ -235,7 +235,9 @@ public static Map<Integer, Object> toMap(Map<Integer, Type.PrimitiveType> idToTy
         idToMetricMap.forEach((id, value) -> {
             Type.PrimitiveType type = idToTypeMapping.get(id);
-            map.put(id, Conversions.fromByteBuffer(type, value));
+            if (type != null) {
```
Just for my edification - When can this happen? Should it be an explicit error instead of ignoring the missing type?

ZacBlanco replied:
This occurs when the `idToMetricMap` contains an entry for a non-primitive type, since the `idToTypeMapping` only contains mappings for primitive types (e.g. we were using a lot of functions with `row` types). I'm not sure the cases when this actually occurs, but I had to put this check in due to some NPEs when running queries with some of the sample tables.

IIRC it may be that this change might not be explicitly required for this PR. I think it got chucked in to this commit by accident. If not included here I will probably include it in another PR down the road.

ClarenceThreepwood replied:
I'd suggest removing for now. And then add it later with a test case that triggers this

ZacBlanco replied:
fixed

agrawalreetika commented:
Hi @ZacBlanco
Wanted to confirm, does this also add support for DELETE if the WHERE clause matches entire partitions?

agrawalreetika commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java:
```diff
@@ -152,6 +162,8 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
 
     protected abstract ConnectorTableMetadata getTableMetadata(ConnectorSession session, SchemaTableName table);
 
+    protected abstract Table getIcebergTable(ConnectorSession session, SchemaTableName schemaTableName);
```
I think if we are introducing  `getIcebergTable` then we can add this in the separate commit and make of use it even in the existing place where `getHiveIcebergTable` or `getNativeIcebergTable` is being called?

ZacBlanco replied:
Sure, I can remove this for now. I will open another PR for the refactor later.

ZacBlanco commented:
> Hi @ZacBlanco Wanted to confirm, Does this also add support for DELETE if the WHERE clause matches entire partitions or do you know if that's already supported?

No, this does not support any kind of `WHERE` clause currently

ChunxuTang commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java:
```diff
@@ -361,7 +361,7 @@ public void testExplainAnalyzeDDL()
     }
 
-    private void assertExplainAnalyze(@Language("SQL") String query)
+    protected void assertExplainAnalyze(@Language("SQL") String query)
```
Why would you need to change it from `private` to `protected`? Could you run the tests without this change?

ZacBlanco replied:
I updated this so I could call this method in the test. e.g. here: https://github.com/prestodb/presto/pull/20776/files#diff-e86735ad8b4b0dd7856263d9696e731dd986026f5547b8f78265e1e389a308daR99

ChunxuTang replied:
I assume you changed the visibility because this PR only partially supports the base `testDelete` function. I don't suggest updating the visibility due to partial support of the deletion feature.
Would you happen to have a plan to support other deletion syntaxes?

ChunxuTang replied:
Because the function was designed to be `private`. If this PR supports the full `testDelete` function, we probably wouldn't need to update the visibility as we could just call the base `testDelete` function, right?

ChunxuTang replied:
That's also why I asked about your plan to support other deletion syntaxes.

ZacBlanco replied:
The plan for supporting the rest of the delete syntax is going to be up to the Presto's iceberg contributors. The next step is probably to support whole-partition deletes, but I don't have the intention of including that in this PR. 

I've changed the visibility back on this method

ChunxuTang reviewed (changes requested):
Hi @ZacBlanco, thanks for the contribution. I left a few comments below.

ZacBlanco commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergNativeMetadata.java:
```diff
@@ -332,4 +340,30 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
     }
+
+    @Override
+    public OptionalLong metadataDelete(ConnectorSession session, ConnectorTableHandle tableHandle, ConnectorTableLayoutHandle tableLayoutHandle)
+    {
+        IcebergTableLayoutHandle layoutHandle = (IcebergTableLayoutHandle) tableLayoutHandle;
+        if (!supportsMetadataDelete(session, tableHandle, Optional.of(layoutHandle))) {
+            throw new PrestoException(INVALID_ARGUMENTS, "iceberg metadata delete doesn't support predicates");
+        }
+        IcebergTableHandle handle = (IcebergTableHandle) tableHandle;
+        org.apache.iceberg.Table icebergTable = getNativeIcebergTable(resourceFactory, session, handle.getSchemaTableName());
+        transaction = icebergTable.newTransaction();
+        DeleteFiles deletes = transaction.newDelete();
+        AtomicLong rowsDeleted = new AtomicLong(0L);
+        try (CloseableIterable<FileScanTask> files = icebergTable.newScan().planFiles()) {
+            files.forEach(t -> {
+                deletes.deleteFile(t.file());
+                rowsDeleted.addAndGet(t.estimatedRowsCount());
+            });
+        }
+        catch (IOException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "failed to scan files for delete", e);
+        }
+        deletes.commit();
+        transaction.commitTransaction();
+        return OptionalLong.of(rowsDeleted.get());
+    }
 }
```
Hey @ChunxuTang, I would have combined these but for now I plan to leave them separate. I have another PR open which actually combines a lot of logic from both implementations. If this PR gets in first I will reduce the implementation into a single method in the other PR: https://github.com/prestodb/presto/pull/20856

ZacBlanco replied:
Originally, there was a single implementation, but changed it due to the review comment here: https://github.com/prestodb/presto/pull/20776#discussion_r1324004195

ChunxuTang replied:
I noticed that https://github.com/prestodb/presto/pull/20856 has been merged. Would you adjust this PR accordingly?

ZacBlanco replied:
This PR has been updated to include a single implementation now 😄

agrawalreetika commented:
@ZacBlanco Can we also add the documentation for this along with few examples and limitations?

ZacBlanco commented:
> @ZacBlanco Can we also add the documentation for this along with a few examples and limitations? And also could you please squash the commit as well?

Fixed. I've also squashed all the commits

steveburnett commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -671,9 +671,38 @@ exists as we've rolled back to the previous state.
     -----------+------+-----------+---------
     (0 rows)
 
+DELETION
```
Looks great! 

One question: 
Would it be appropriate to consider naming this new content DELETE, and move it from the end of the page into the SQL Support heading as a subheading between ALTER TABLE and DROP TABLE?

ZacBlanco replied:
Thanks for the quick review! I agree that your suggestion for the placement of the section is better. I've fixed it.

steveburnett reviewed (approved):
LGTM! (docs)

agrawalreetika reviewed (approved):
LGTM 👍🏻  Please review the test failures.

ChunxuTang commented:
@ZacBlanco Thanks for your work and the clarification on your plan to support deletion.

Hi @tdcmeehan and @beinan, would you mind sharing your thoughts on this PR? It represents an initial step towards adding support for deletion in the Iceberg connector. However, from the discussion with @ZacBlanco and from my knowledge, it appears that we currently don't have a concrete plan to support the complete set of deletion syntax, but rather a preliminary feature that covers "DELETE FROM <table>" statements without predicates. This means that customers may have access to an incomplete set of deletion syntax for an extended period.

I'm unsure whether we should proceed with delivering this partial deletion feature, allowing customers to use at least the basic functionality, or if it would be better to postpone the delivery until we have a clearer roadmap for full deletion syntax support. Any suggestions will be appreciated.

tdcmeehan commented:
Why not simply implement `TRUNCATE` rather than `DELETE`?

tdcmeehan commented:
I think it's worthwhile to add support for TRUNCATE, since that's semantically more clear.

@ChunxuTang : I think a logical series of steps would be first support truncation, then deletion by partition (works in v1 and v2), then row level deletion (works only with v2 tables).  We can add issues and track in our Iceberg project board the latter two issues.  WDYT?

ZacBlanco commented:
The commit, description, and PR title have all been updated to indicate truncate is implemented instead

tdcmeehan commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -674,6 +701,6 @@ exists as we've rolled back to the previous state.
 Iceberg Connector Limitations
 -----------------------------
 
-* :doc:`/sql/delete` is only supported if the ``WHERE`` clause matches entire partitions.
```
I suppose the old text was just wrong?

ZacBlanco replied:
I just fixed this now -- but yes it was wrong

tdcmeehan commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -479,6 +479,30 @@ Alter table operations are supported in the connector::
+
+    TRUNCATE TABLE nation;
+
+.. code-block:: text
```
```suggestion
.. code-block:: sql
```

ZacBlanco replied:
This is actually the output from the CLI after running `TRUNCATE TABLE nation`, so I think keeping `text` here is ok.

tdcmeehan replied:
Ah, got it!

steveburnett commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -479,6 +479,30 @@ Alter table operations are supported in the connector::
+TRUNCATE
+^^^^^^^^
+
+The iceberg connector can delete all of the data data from tables without
```
```suggestion
The iceberg connector can delete all of the data from tables without
```

ZacBlanco replied:
Fixed, thanks for pointing out the typo!

steveburnett reviewed (changes requested):
Just one thing - "data data" seems unintentional. Yes?

steveburnett reviewed (approved):
LGTM! (docs)

New build of doc, typo mentioned in previous review verified as fixed. Thanks @ZacBlanco!

tdcmeehan commented:
@ChunxuTang / @beinan any further concerns?

ChunxuTang commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java:
```diff
@@ -434,4 +434,36 @@ public Optional<SystemTable> getSystemTable(ConnectorSession session, SchemaTabl
         return getIcebergSystemTable(tableName, icebergTable);
     }
+
+    /**
+     * Deletes the files from a particular scan
+     *
+     * @return
+     */
```
The comment here looks unnecessary. Could you remove it?

ZacBlanco replied:
I kept the comment but updated it to properly reflect the functionality

ChunxuTang commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java:
```diff
@@ -434,4 +434,36 @@ public Optional<SystemTable> getSystemTable(ConnectorSession session, SchemaTabl
+     *
+     * @return
+     */
+    private long deleteTable(Table icebergTable, Iterable<FileScanTask> scan)
```
With the new `truncateTable` function, the name of this function (`deleteTable`) looks a bit misleading. How about `deleteFiles`? Or any other ideas?

ZacBlanco replied:
I've updated the name to `removeScanFiles`

ChunxuTang reviewed (commented):
@tdcmeehan Thanks for your suggestion. Renaming the commit to support `TRUNCATE` looks nice to me.

@ZacBlanco Thanks for your consistent and nice contributions. I left a few comments. Would you mind taking a look?

ChunxuTang commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java:
```diff
@@ -277,12 +283,6 @@ public ColumnHandle getUpdateRowIdColumnHandle(ConnectorSession session, Connect
         return primitiveIcebergColumnHandle(0, "$row_id", BIGINT, Optional.empty());
     }
 
-    @Override
-    public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle)
-    {
-        throw new PrestoException(NOT_SUPPORTED, "This connector only supports delete where one or more partitions are deleted entirely");
-    }
-
```
Quick question: As this PR turns to focus on supporting `TRUNCATE`, do we still need to remove this function?

ZacBlanco replied:
This method is incorrectly implemented and misdirects users. If someone were to query the connector `DELETE from <table> WHERE <predicate>;` the error would say "This connector only supports delete where one or more partitions are deleted entirely".

However, that statement is incorrect and misdirects users as the connector never supported deleting partitions. This should be removed until the support is added.

ChunxuTang reviewed (approved):
@ZacBlanco Thanks for the work.

Generally lgtm. Just left a minor comment.
